File: sorting/quicksort.test.js
const {quicksort, partition} = require('./quicksort')

test('partition of quicksort', () => {
    const arr = [3, 4, 1, 6, -1]
    const pivotIndex = partition(arr, 0, arr.length)
    expect(pivotIndex).toEqual(2)
    expect(arr).toEqual([-1, 1, 3, 6, 4])
})

test('quick sort full', () => {
    const arr = [3, 2, 6, 1, 4, 6, 100]
    quicksort(arr)
    const expected = [1, 2, 3, 4, 6, 6, 100]
    expect(arr).toEqual(expected)
})

test('quick sort empty', () => {
    const arr = []
    quicksort(arr)
    const expected = []
    expect(arr).toEqual(expected)
})

test('quick sort minimal', () => {
    const arr = [2, 1]
    quicksort(arr)
    const expected = [1, 2]
    expect(arr).toEqual(expected)
})

File: sorting/quicksort_test.go
package main

import (
	"reflect"
	"testing"
)

func TestPartition(t *testing.T) {
	arr := []int{13, 5, 1, 7, 10, 14, 100}
	expectedPartion := []int{10, 5, 1, 7, 13, 14, 100}
	expectedPivotIndex := 4
	givenPivotIndex := Partition(arr, 0, len(arr))
	if givenPivotIndex != expectedPivotIndex {
		t.Errorf("Given index '%v' doesn't equal to expected '%v'", givenPivotIndex, expectedPivotIndex)
	}
	if !reflect.DeepEqual(arr, expectedPartion) {
		t.Errorf("Given arr '%v' doesn't equal to expected '%v'", arr, expectedPartion)
	}
}

func TestQuicksort(t *testing.T) {
	arr := []int{13, 5, 1, 7, 10, 14, 100}
	expected := []int{1, 5, 7, 10, 13, 14, 100}
	Quicksort(arr)
	if !reflect.DeepEqual(arr, expected) {
		t.Errorf("Given arr '%v' doesn't equal to expected '%v'", arr, expected)
	}
}
